crates/transport/src/layers/retry.rs

```diff
@@ -0,0 +1,232 @@
+use crate::{
+    error::{TransportError, TransportErrorKind},
+    RpcErrorExt, TransportFut,
+};
+use alloy_json_rpc::{RequestPacket, ResponsePacket};
+use std::{
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+use tracing::trace;
+
+/// A Transport Layer that is responsible for retrying requests based on the
+/// error type. See [`TransportError`].
+///
+/// TransportError: crate::error::TransportError
+#[derive(Debug, Clone)]
+pub struct RetryBackoffLayer {
+    /// The maximum number of retries for rate limit errors
+    max_rate_limit_retries: u32,
+    /// The maximum number of retries for timeout errors
+    max_timeout_retries: u32,
+    /// The initial backoff in milliseconds
+    initial_backoff: u64,
+    /// The number of compute units per second for this provider
+    compute_units_per_second: u64,
+}
+
+impl RetryBackoffLayer {
+    /// Creates a new retry layer with the given parameters.
+    pub const fn new(
+        max_rate_limit_retries: u32,
+        max_timeout_retries: u32,
+        initial_backoff: u64,
+        compute_units_per_second: u64,
+    ) -> Self {
+        Self {
+            max_rate_limit_retries,
+            max_timeout_retries,
+            initial_backoff,
+            compute_units_per_second,
+        }
+    }
+}
+
+/// [RateLimitRetryPolicy] implements [RetryPolicy] to determine whether to retry depending on the
+/// err.
+#[derive(Debug, Copy, Clone, Default)]
+#[non_exhaustive]
+pub struct RateLimitRetryPolicy;
+
+/// [RetryPolicy] defines logic for which [TransportError] instances should
+/// the client retry the request and try to recover from.
+pub trait RetryPolicy: Send + Sync + std::fmt::Debug {
+    /// Whether to retry the request based on the given `error`
+    fn should_retry(&self, error: &TransportError) -> bool;
+
+    /// Providers may include the `backoff` in the error response directly
+    fn backoff_hint(&self, error: &TransportError) -> Option<std::time::Duration>;
```

Comment:
this trait seems redundant/ should be merged with the other error trait

Reply:
this is indeed redundant and was already redundant in the foundry code this is based on, the original idea was to plug this into the retrybackofflayer but we don't use this trait, I'd like to keep this because we could replace the RateLimitRetryPolicy with a box<dyn or smth